lstchain/tools/lstchain_dl1_to_dl2.py

```diff
@@ -38,44 +38,131 @@
+    -----------
+    dl1_filename : str
+        Name of the DL1 file
+
+    Returns:
+    --------
+    str
+        Name of the DL2 file
+    """
+    return dl1_filename.replace('dl1', 'dl2', 1)
+
+class DL1ToDL2Tool(Tool):
+    name = "DL1 to DL2 Tool"
+    description = __doc__
+
+    input_files = traits.List(
+        traits.Path,
+        help="Path (or list of paths) to a DL1 HDF5 file",
+    ).tag(config=True)
+
+    path_models = traits.Path(
+        help="Path where to find the trained RF",
+        default_value='./trained_models',
+    ).tag(config=True)
+
+    output_dir = traits.Path(
+        help="Path where to store the reco dl2 events",
+        default_value='./dl2_data',
+    ).tag(config=True)
+
+    config_file = traits.Path(
+        allow_none=True,
+        help="Path to a configuration file. If none is given, a standard configuration is applied",
+        default_value=None,
+    ).tag(config=True)
+
+    aliases = {
+        ("f", "input-files"): "DL1ToDL2Tool.input_files",
+        ("p", "path-models"): "DL1ToDL2Tool.path_models",
+        ("o", "output-dir"): "DL1ToDL2Tool.output_dir",
+        ("c", "config"): "DL1ToDL2Tool.config_file",
+    }
+
+    def setup(self):
+         
+        # Check if input files are provided
+        if not self.input_files:
+            raise ToolConfigurationError("No input files provided. Use --input-files to specify.")
+
+        # Additional setup logic can go here
+        self.log.info(f"Input files: {self.input_files}")
+        self.log.info(f"Path to models: {self.path_models}")
+        self.log.info(f"Output directory: {self.output_dir}")
+
+    def start(self):
+            
+        custom_config = {}
+        if self.config_file is not None:
+            try:
+                custom_config = read_configuration_file(self.config_file.absolute())
+            except Exception as e:
+                self.log.error(f"Custom configuration could not be loaded: {e}")
+                return
+
+        config = replace_config(standard_config, custom_config)
+
+        models_keys = ['reg_energy', 'cls_gh']
+
+        if config['disp_method'] == 'disp_vector':
+            models_keys.append('reg_disp_vector')
+        elif config['disp_method'] == 'disp_norm_sign':
+            models_keys.extend(['reg_disp_norm', 'cls_disp_sign'])
+
+        models_dict = {}
+        for models_key in models_keys:
+            models_path = Path(self.path_models, f'{models_key}.sav')
+
+            if len(self.input_files) == 1:
+                models_dict[models_key] = models_path
+            else:
+                models_dict[models_key] = joblib.load(models_path)
+
+        for input_dl1file in self.input_files:
+            output_filepath = apply_to_file(input_dl1file, models_dict, self.output_dir, config, 
+                                            self.path_models, self.overwrite)
+            p = Provenance()
+            p.add_input_file(input_dl1file, role='dl1 input file')
```

Comment:
To make sure provenance is also correct when errors happen, you should add input / output files as close as possible to opening them.